internal/qldb/account.go

```diff
@@ -0,0 +1,171 @@
+	_, err := Repo.Driver.Execute(ctx,
+		func(txn qldbdriver.Transaction) (interface{}, error) {
+			accountId := model.GenerateAccountId(userId, currency)
+			result, err := txn.Execute(`
+            SELECT * 
+            FROM Accounts 
+            WHERE id = ?`, accountId)
+			if err != nil {
+				return nil, err
+			}
+			if result.Next(txn) {
+				return nil, nil
+			}
+
+			initialDecimal := ion.NewDecimal(initialBalance, 0, false)
+			a := &model.QldbAccount{
+				Id:          model.GenerateAccountId(userId, currency),
+				Currency:    currency,
+				UserId:      userId,
+				Balance:     initialDecimal,
+				Hold:        ion.MustParseDecimal("0"),
+				Available:   initialDecimal,
+				UpdatedAt:   time.Now(),
+				AccountUUID: uuid.New().String(),
+			}
+
+			_, err = txn.Execute("INSERT INTO Accounts ?", a)
+
+			return nil, err
+		})
+
+	return err
+}
+
+func getAccountQldbTransaction(
+	txn qldbdriver.Transaction,
+	accountId string) (*model.QldbAccount, error) {
+	var account *model.QldbAccount
+	result, err := txn.Execute(`
+        SELECT * 
+        FROM Accounts 
+        WHERE id = ?`, accountId)
+	if err != nil {
+		return nil, err
+	}
+	if result.Next(txn) {
+		if err := ion.Unmarshal(
+			result.GetCurrentData(), &account); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, errors.New(AccountNotFoundError)
+	}
+    if result.Err() != nil {
+        return nil, err
+    }
+	return account, nil
+}
+
+func creditAccountUpdate(
+    txn qldbdriver.Transaction,
+    account *model.QldbAccount,
+    amount *big.Int) error {
+	balance, _ := account.Balance.CoEx()
+	balance.Add(balance, amount)
+	hold, _ := account.Hold.CoEx()
+	available, _ := account.Available.CoEx()
+	available.Sub(balance, hold)
+
+	_, err := txn.Execute(`
+        UPDATE Accounts AS a 
+        SET a.balance = ?, a.available = ?, a.updatedAt = ? 
+        WHERE a.id = ?`,
+		ion.NewDecimal(balance, 0, false),
+		ion.NewDecimal(available, 0, false),
+		time.Now(),
+		account.Id)
+	return err
+}
+
+func debitAccountUpdate(
+    txn qldbdriver.Transaction,
+    account *model.QldbAccount,
+    amount *big.Int,
+    commission *big.Int,
+    feeAmount *big.Int) error {
+	hold, _ := account.Hold.CoEx()
+	hold.Sub(hold, amount)
+	hold.Sub(hold, commission)
+	hold.Sub(hold, feeAmount)
+	balance, _ := account.Balance.CoEx()
+	balance.Sub(balance, amount)
+	balance.Sub(balance, commission)
+	balance.Sub(balance, feeAmount)
+	available := new(big.Int)
+	available.Sub(balance, hold)
+
+	_, err := txn.Execute(`
+        UPDATE Accounts AS a 
+        SET a.balance = ?, a.hold = ?, a.available = ?, a.updatedAt = ? 
+        WHERE a.id = ?`,
+		ion.NewDecimal(balance, 0, false),
+		ion.NewDecimal(hold, 0, false),
+		ion.NewDecimal(available, 0, false),
+		time.Now(),
+		account.Id)
+	return err
+}
+
+func holdBalanceUpdate(
+	txn qldbdriver.Transaction,
+	account *model.QldbAccount,
+	amount *big.Int,
+	isSub bool) error {
+	hold, _ := account.Hold.CoEx()
```

Comment:
What's the second param? Is that an error?

Reply:
The second parameter is the exponent. I was swallowing it as we set everything as integers. Alternatively I could wrap these calls in a function that returns an error since if that exponent is not 1 then we know it's incorrect